app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java

```diff
@@ -0,0 +1,432 @@
+
+        /**
+         *
+         * @param jsonReport JSONObject containing tests data
+         * @param test element of a test - must contain bitmap and ingredients fields
+         * @param countDownLatch used to signal the task completion
+         * @param semaphore semaphore used to signal the end of the task
+         */
+        public RunnableTest(JSONObject jsonReport, TestElement test, CountDownLatch countDownLatch, Semaphore semaphore) {
+            this.jsonReport = jsonReport;
+            this.test = test;
+            this.countDownLatch = countDownLatch;
+            this.semaphore = semaphore;
+        }
+
+        @Override
+        public void run() {
+            try {
+
+                Log.d(TAG,"RunnableTest -> id \"" + Thread.currentThread().getId() + "\" started");
+                long started = java.lang.System.currentTimeMillis();
+
+                //evaluate text extraction confidence
+                String extractedIngredients = executeOcr(test.getPicture());
+                String correctIngredients = test.getIngredients();
+                float confidence = ingredientsTextComparison(correctIngredients, extractedIngredients);
+
+                //insert test in report
+                test.setConfidence(confidence);
+                //insert extracted test
+                test.setRecognizedText(extractedIngredients);
+
+
+                addTestElement(jsonReport, test);
+
+                //done test process signal
+                countDownLatch.countDown();
+
+                long ended = java.lang.System.currentTimeMillis();
+                Log.d(TAG,"RunnableTest -> id \"" + Thread.currentThread().getId() + "\" ended (runned for " + (ended - started) + " ms)");
+
+                //let start another task
+                semaphore.release();
+
+            }catch (JSONException e) {
+                Log.e(TAG, "Error elaborating JSON test element");
+            }
+        }
+    }
+
+    /**
+     * Puts the json object of the TestElement inside the JSONObject jsonReport, multi thread safe
+     * @param jsonReport the report containing tests in JSON format
+     * @param test element of a test
+     * @throws JSONException
+     * @author Luca Moroldo (g3)
+     */
+    synchronized void addTestElement(JSONObject jsonReport, TestElement test) throws JSONException {
+        jsonReport.put(test.getFileName(), test.getJsonObject());
+    }
+
+    /**
+    * Returns a HashMap of (Tag, Value) pairs where value is the average test result of the photos tagged with that Tag
+    * @author Nicolò Cervo (g3) with the tutoring of Francesco Pham (g3)
+    */
+    public HashMap getTagsStats() throws JSONException {
+        HashMap<String, Float> tagStats = new HashMap<>(); //contains the cumulative score of every tag
+        HashMap<String, Integer> tagOccurrences = new HashMap<>(); //contains the number of occurrences of each tag
+        for(TestElement element : testElements) {
```

Comment:
In the declaration of the variables you can add a comment to explain what testElements contains